src/libraries/Microsoft.Extensions.Configuration.Binder/gen/ConfigurationBindingGenerator.Parser.cs

```diff
@@ -140,18 +140,19 @@ private void ProcessBindCall(BinderInvocationOperation binderOperation)
                 }
 
                 ITypeSymbol? type = ResolveType(objectArg.Value)?.WithNullableAnnotation(NullableAnnotation.None);
-                INamedTypeSymbol? namedType;
 
-                if ((namedType = type as INamedTypeSymbol) is null ||
-                    namedType.SpecialType == SpecialType.System_Object ||
-                    namedType.SpecialType == SpecialType.System_Void ||
-                    // Binding to root-level struct is a no-op.
-                    namedType.IsValueType)
+                if (!ValidateRootConfigType(overload, type, binderOperation.Location))
                 {
                     return;
                 }
 
-                AddRootConfigType(methodGroup: BinderMethodSpecifier.Bind, overload, namedType, binderOperation.Location);
+                if (type.IsValueType)
+                {
+                    _context.ReportDiagnostic(Diagnostic.Create(Diagnostics.ValueTypesInvalidForBind, binderOperation.Location, type));
```

Comment:
Nit: adding a dependency on `SourceProductionContext` in the parser means that its outputs can never be cached as incremental. You should consider storing the `Diagnostic` instances in an a temporary list to be incorporated in the model that gets passed down to the emitter.